### ISSUE TITLE
Accessing Global Variables in a Class

### ISSUE THREAD

**Issue by APuente21** [susanBuck/dwa15-fall2017#41]
I don't have an issue, at least not anymore, but I thought I would post this for the benefit of the class (and my progress log :)

In PHP, if you have an object with some global parameters, and you want to use these parameters in your class functions, you need to use the `$this->` notation. This is a bit different from other languages, like Java and C++, where you don't have to do it, unless you have the same parameter already defined locally. 

Take the bellow code as an example, in PHP if you omit the `$this->` from the `getA()` method, you would probably get a `variable undefined error`, in Java and C++ you wouldn't. I found this out the hard way when I spent about 15min trying to figure out why I was getting that error message

```
class Foo
{
    public $a = "I'm a!";

    public function getA() {
        return $this->a;
    }
}
```

**Comment by ysrinu** [comment]
@APuente21 , by "some global parameters" did you mean $a here?
It's a variable here and $this works the same way as 'this' in C#,C++,Java and 'me' in VB.
"this" simply means the current instance of the class.

For example
```
class Foo
{
    int a=10;

    void function getA(int a=20) {
        print(a); // prints '20'
        print(this.a); //prints 10
    }
}
```
Or you might be referring that for a variable defined in the class (and not as function parameter or local variable) has to be accessed with $this, that makes sense too. Even in that case this stands for the current instance of the class instead of 'global'.
To refer global variables you should use the keyword 'global' or alternatively use the $GLOBALS array
http://php.net/manual/en/language.variables.scope.php

Bottom line, 'this' doesn't mean global. It means the current instance of the class and by that definition it refers to variable defined as class level which happens to be in global scope in comparsion to the local scope of your function.  Nevertheless, 'this' purpose is not about local versus global.

**Comment by APuente21** [comment]
I'm referring to the variable defined in the class. In php, if you are within a class method, and the class already has some defined variables, say through the function __ construct() method, you need to use the  `$this->`  notation to access the variable. If you modify your code to what I show below, this would return a undefined variable exception, in Java and C++ it would print 10;

class Foo
{
    int a=10;

    void function getA() {
        print(a); // undefined variable
    }
}

**Comment by ysrinu** [comment]
@APuente21 , agree but again that's a difference in the scope resolution between the languages.
In C++, if the variable is not defined in the function parameter or within function it will look at the 'parent' scope and find it. 
In php it won't find global variable unless you defined with 'global' keyword first
OR
you can access class members by using 'this' keyword, which works the same in C++ too.
It's easy to misunderstand that 'this' and 'global' are the same, the effect could be similar but there's a subtle difference.
'this' means and always means the current instance of this class and you are using that here to refer to the variables defined at the class level (which also happen to be at the global level).
